Logout, and some cleaning

ChickenProp opened the pull request:
Adds ability to logout from D-cookie-logons, and mostly switches to qualified imports in A, B and D (I didn't bother with C).

ChickenProp committed:
Use qualified imports in A-servant-start.

I'm not sure what's going on with (:<|>). If I try to use it unqualified
in the definition of `server`, it suggests that I might want to include
it in either the "import Servant (...)" or "import Servant.Api (...)"
lines, even though it's already there (and even if I put it in the
Servant line as well).

ChickenProp committed:
Use qualiied imports in B-postgres-data.

Same confusion as previous commit re :<|>.

ChickenProp committed:
Matthew figured out the :<|> thing.

"I looked this up a little more thoroughly. the (:<|>) operator is
defined here:
https://github.com/haskell-servant/servant/blob/master/servant/src/Servant/API/Alternative.hs
it's got two instances, one for a Monoid and one for a Semigroup. I
guess we need the (..) to ensure we get both."

https://gist.github.com/mebassett/08e0cbe81187b0e5ff94c3df192e53b5

ChickenProp committed:
Add logout ability to D.

This also switches to mostly-qualified imports, it's a messy commit.

Bug: you can no longer hit "enter" to login. That's because the button
is no longer a submit button. As a submit button, it refreshed the page,
which isn't really what we'd want I think? But hitting "enter" to login
would be pretty much an essential feature in a real app.

(You can admittedly use "enter" if you tab to the login button.)

Bug: logs don't disappear if you logout. That's because we never reset
model.logs. We could do that on logout, but resetting every piece of
state separately seems like not what we want. I'd suggest a better way
to go might be to put all the login-protected info in a type, and give
model a Maybe one-of-those.

trippingtarballs commented:
lgtm 👍 I cannot say I fully understand all that's going on. And you'll have to explain to me again the merits of `import` techniques in Haskell.

ChickenProp committed:
Remove outdated comment

I added it when I needed to qualify the uses of :<|> in server, and then forgot to remove it when I no longer needed to qualify those.

mebassett commented:
this import technique is makes it easier to see what you're using and where things come from.  It's a pythonic style (that I haven't seen used a lot when reading Haskell) that I very much prefer.   

@ChickenProp 's efforts on this are also a good learning experience, cause you have to figure out what comes from what.  I tried to use a minimal set of language extensions but was too lazy to do minimal imports.  You'll notice the official servant-examples use quite a few of both.  

This reduces cargo-culting.
